### PR TITLE
namespaceChecker

### PR DESCRIPTION
**Pull request by dantin**
Add namespaceChecker to move peer to the targeted store

### CODE REVIEW

File: server/core/tablecodec.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package core
```
Comment (disksing) [inline server/core/tablecodec.go]: I don't think we should include tablecodec.go in this PR.

File: server/schedule/namespace_checker.go [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+}
+
+// Check verifies a region's namespace, creating an Operator if need.
+func (n *NamespaceChecker) Check(region *core.RegionInfo) *Operator {
+	// fail-fast if there is only ONE namespace
+	if n.classifier == nil || len(n.classifier.GetAllNamespaces()) == 1 {
+		return nil
+	}
+	for _, peer := range region.GetPeers() {
+		newPeer := n.SelectBestPeerToRelocate(region, n.filters...)
```
Comment (disksing) [inline server/schedule/namespace_checker.go]: I think we need to check if `peer` is located on a store that is already belongs to the right namespace.

File: server/core/tablecodec.go (hunk 1 above)
Comment (dantin) [reply server/core/tablecodec.go]: Already moved to PR761

File: server/schedule/namespace_checker.go (hunk 2 above)
Comment (dantin) [reply server/schedule/namespace_checker.go]: change logic which checking the `peer` location first

File: server/namespace_test.go [hunk 3]
```diff
@@ -63,6 +63,33 @@ func (s *testNamespaceSuite) TestReplica(c *C) {
...
+	s.classifier.setStore(2, "ns1")
+	s.classifier.setStore(3, "ns2")
+
+	checker := schedule.NewNamespaceChecker(s.opt, s.tc, s.classifier)
+
+	// Namespace should be added to the store with the same namespace.
+	s.classifier.setRegion(1, "ns2")
+	s.tc.addLeaderRegion(1, 1)
+	op := checker.Check(s.tc.GetRegion(1))
+	checkTransferPeer(c, op, 1, 3)
```
Comment (disksing) [inline server/namespace_test.go]: I suppose we can add a test case that some peers of a region is in the right namespace, while some are not. The checker should select the ones in the wrong namespace to move.

**Comment by disksing** [comment]
The reset LGTM. PTAL @nolouch @siddontang

File: server/namespace_test.go (hunk 3 above)
Comment (dantin) [reply server/namespace_test.go]: Done

**Comment by disksing** [comment]
LGTM.

File: server/schedule/namespace_checker.go [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+	// get all the stores belong to the namespace
+	targetStores := n.getNamespaceStores(region)
+	if len(targetStores) == 0 {
+		return nil
+	}
+	for _, peer := range region.GetPeers() {
+		// check whether the peer has been already located on a store that is belong to the target namespace
+		if n.isExists(targetStores, peer.StoreId) {
+			continue
```
Comment (nolouch) [inline server/schedule/namespace_checker.go]: you can directly use Namespace filter in SelectBestPeerToRelocate  function

File: server/namespace_test.go [hunk 5]
```diff
@@ -63,6 +63,49 @@ func (s *testNamespaceSuite) TestReplica(c *C) {
...
+	s.classifier.setRegion(2, "ns1")
+	s.tc.addLeaderRegion(2, 1)
+	s.classifier.setRegion(3, "ns2")
+	s.tc.addLeaderRegion(3, 2)
+	op = checker.Check(s.tc.GetRegion(2))
+	c.Assert(op, IsNil)
+	op = checker.Check(s.tc.GetRegion(3))
+	checkTransferPeer(c, op, 2, 3)
+
+	// Stop move region if namespace is added in the same namespace.
```
Comment (siddontang) [inline server/namespace_test.go]: > namespace is added in the same namespace

I guess the comment will confuse the user, what is the same namespace?

File: server/schedule/namespace_checker.go (hunk 4 above)
Comment (dantin) [reply server/schedule/namespace_checker.go]: I don't think it's a good idea to call `getNamespaceStores()` every time in the loop.

File: server/namespace_test.go (hunk 5 above)
Comment (dantin) [reply server/namespace_test.go]: Fixed

File: server/namespace_test.go [hunk 6]
```diff
@@ -63,6 +63,49 @@ func (s *testNamespaceSuite) TestReplica(c *C) {
...
+	s.tc.addRegionStore(1, 0)
+	s.tc.addRegionStore(2, 10)
+	s.tc.addRegionStore(3, 0)
+	s.classifier.setStore(1, "ns1")
+	s.classifier.setStore(2, "ns1")
+	s.classifier.setStore(3, "ns2")
+
+	checker := schedule.NewNamespaceChecker(s.opt, s.tc, s.classifier)
+
+	// Namespace should be added to the store with the same namespace.
```
Comment (siddontang) [inline server/namespace_test.go]: the comment confuses the user too here.

Comment (dantin) [reply server/namespace_test.go]: Fixed

File: server/schedule/namespace_checker.go (hunk 4 above)
Comment (nolouch) [reply server/schedule/namespace_checker.go]: got it

Comment (dantin) [reply server/schedule/namespace_checker.go]: thx

**Pull request merged**